### PR TITLE
Dont fuse write into read if indexing differs

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118210

Fix for https://github.com/pytorch/pytorch/issues/101950, https://github.com/pytorch/pytorch/issues/94693

Similar to inplacing a kernel only fuse a write after a read of the same tensor if the write and read have same indexing formula. I did a perf test and it was neutral.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by mlazos** [comment]
Why doesn't https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_inductor/scheduler.py#L2004C9-L2019C42
cover this case?

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -2030,6 +2030,25 @@ def can_fuse_vertical(self, node1, node2):
             if node1_names & self.name_to_fused_node[name].ancestors:
                 why("intermediate nodes between node1 & node2")
                 return False
+
+        # similar to can_inplace, if we are going to fuse a write subsequent to a read
+        # require that the indexing and size is the same
+        for write in node2.read_writes.writes:
```
Comment (masnesral) [inline torch/_inductor/scheduler.py:2036]: Does it matter for perf that this is n^2? I have no context on how many iterations we're talking about here.

**Review by masnesral — COMMENTED** [review]
LGTM. Lazos had a question, so I'll let him approve.

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/scheduler.py:2036]: fair question, this is using same pattern already in this function but I can refactor both to make them O(n) (no compilation increase showed up in benchmark)

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -2030,6 +2030,25 @@ def can_fuse_vertical(self, node1, node2):
...
+            for read in node1.read_writes.reads:
+                if write.name != self.mutation_renames.get(read.name, read.name):
+                    continue
+
```
Comment (eellison) [inline torch/_inductor/scheduler.py:2040]: @mlazos that was checking producer reads with consumer writes where as this is checking produce writes with consumer reads but I can refactor the check and reuse it

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (masnesral) [reply torch/_inductor/scheduler.py:2036]: I defer to you and whether it's worth it to refactor

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**